parser/src/parser/tests/variables.rs

```diff
@@ -1,179 +1,130 @@
-use super::{Identifier, IntegrityConstraint, ParseTest, Source, SourceSection};
-use crate::ast::{
-    AccessType, ConstraintExpr, Expression::*, InlineConstraintExpr, IntegrityStmt::*,
-    SymbolAccess, VariableBinding, VariableValueExpr,
-};
+use miden_diagnostics::{SourceSpan, Span};
+
+use crate::ast::*;
+
+use super::ParseTest;
 
 // VARIABLES
 // ================================================================================================
 #[test]
 fn variables_with_and_operators() {
     let source = "
-    integrity_constraints:
+    mod test
+
+    ev test([clk]):
         let flag = n1 & !n2
         enf clk' = clk + 1 when flag";
-    let expected = Source(vec![SourceSection::IntegrityConstraints(vec![
-        VariableBinding(VariableBinding::new(
-            Identifier("flag".to_string()),
-            VariableValueExpr::Scalar(Mul(
-                Box::new(SymbolAccess(SymbolAccess::new(
-                    Identifier("n1".to_string()),
-                    AccessType::Default,
-                    0,
-                ))),
-                Box::new(Sub(
-                    Box::new(Const(1)),
-                    Box::new(SymbolAccess(SymbolAccess::new(
-                        Identifier("n2".to_string()),
-                        AccessType::Default,
-                        0,
-                    ))),
-                )),
-            )),
-        )),
-        Constraint(IntegrityConstraint::new(
-            ConstraintExpr::Inline(InlineConstraintExpr::new(
-                SymbolAccess(SymbolAccess::new(
-                    Identifier("clk".to_string()),
-                    AccessType::Default,
-                    1,
-                )),
-                Add(
-                    Box::new(SymbolAccess(SymbolAccess::new(
-                        Identifier("clk".to_string()),
-                        AccessType::Default,
-                        0,
-                    ))),
-                    Box::new(Const(1)),
-                ),
-            )),
-            None,
-            Some(SymbolAccess(SymbolAccess::new(
-                Identifier("flag".to_string()),
-                AccessType::Default,
-                0,
-            ))),
-        )),
-    ])]);
-    ParseTest::new().expect_ast(source, expected);
+
+    let mut expected = Module::new(ModuleType::Library, SourceSpan::UNKNOWN, ident!(test));
+    // The constraint is converted into a comprehension constraint by the parser, which
+    // involves generating an iterable with one element and giving it a generated binding
+    let body = vec![let_!(flag = expr!(band!(access!(n1), not!(access!(n2)))) =>
+                  enforce_all!(lc!((("%1", range!(0..1))) => eq!(access!(clk, 1), add!(access!(clk), int!(1))), when access!(flag))))];
```

Comment:
Here's what that code would look like macro-expanded:

```rust
Statement::Let(Let::new(
    SourceSpan::UNKNOWN, 
    ident!(flag), 
    expr!(band!(access!(n1), not!(access!(n2)))),
    enforce_all!(ListComprehension::new(
        SourceSpan::UNKNOWN, 
        eq!(access!(clk, 1), add!(access!(clk), int!(1))), 
        vec![(ident!("%1"), range!(0..1))], 
        Some(access!(flag))
    ))
)
```

It isn't too bad in this case, as there is only one `let` and one comprehension; and the comprehension only binds a single iterable which is a constant, but once you get into nested `let` statements, or complex comprehensions, it gets noisy real quick.

Overall, the benefit of the macros is three-fold - making it easy to write assertions against the AST by using macros that approximate the syntax (to the degree possible using macros-by-example); ensuring that changes to the AST do not require updating hundreds of tests (instead only the macro definition need be updated); and reducing the amount of unnecessary noise that instantiating the AST structs typically involves, putting the focus on the syntax structure being defined.

The `let_` and `lc` macros in particular are effective at reducing a lot of noise, as they are the most complicated structures in the AST. That said, it's all tradeoffs - it would obviously be preferable to be able to match the actual language syntax better, as macros-by-example have some odd quirks and limitations, and there isn't a great way to format deeply nested macros, so it can be a bit dense when reading them. On the other hand, the macros here make the tests visible on a single page of code, which means you can see the code being parsed and the macros corresponding to the expected AST at the same time, which is the biggest benefit in my mind, and go a long way to making the macro mess more obvious/clear. I'm open to removing/changing macros which are more of a problem than a help though.

Reply:
Yep, I can imagine how messy it will get with blocks of multiple let expressions, even e.g. just an evaluator with 4-5 `let` and constraint expressions. I'm ok with keeping the macros, but we should see what @bobbinth thinks